service/storage_service.cc

```diff
@@ -6176,6 +6176,162 @@ static bool increases_replicas_per_rack(const locator::topology& topology, const
     return m[dst_rack] + 1 > max;
 }
 
+future<service::group0_guard> storage_service::wait_for_tablet_migration() {
+    auto guard = co_await _group0->client().start_operation(_group0_as, raft_timeout{});
+
+    while (_topology_state_machine._topology.is_busy()) {
+        const auto tstate = *_topology_state_machine._topology.tstate;
+        if (tstate == topology::transition_state::tablet_draining ||
+            tstate == topology::transition_state::tablet_migration) {
+            break;
+        }
+        rtlogger.debug("wait_for_tablet_migration(): topology state machine is busy: {}", tstate);
+        release_guard(std::move(guard));
+        co_await _topology_state_machine.event.wait();
+        guard = co_await _group0->client().start_operation(_group0_as, raft_timeout{});
+    }
+
+    co_return guard;
+}
+
+future<bool> storage_service::exec_tablet_migration(service::group0_guard guard, std::vector<canonical_mutation> updates, sstring reason) {
+    rtlogger.info("{}", reason);
+    rtlogger.trace("do update {} reason {}", updates, reason);
+    updates.emplace_back(topology_mutation_builder(guard.write_timestamp())
+        .set_transition_state(topology::transition_state::tablet_migration)
```

Comment:
This is a known and existing issue move_tablet which this code followed. We talked about this before. We need a solution for both. 

In storage_service::add_repair_tablet_request, we do not modify the tablet map here. Can we modify the tmap directly?The add_repair_tablet_request could be run on any of the node in the cluster.  Is this safe?  How can we do that?

Reply:
We should not introduce new bugs though. 

And I argued that we don't need set_transition_state() at all, so win-win. You don't put tablets into transition, so we don't have to put the state machine in tablet-transition-handling state. idle state is fine, scheduler will start the transitions.

Reply:
> The add_repair_tablet_request could be run on any of the node in the cluster. Is this safe?

It is safe, because the update is executed atomically with the checks by the means of raft. If update is successful, it means the coordinator didn't start transitions concurrently.

Reply:
So, we only execute the updates mutations and skip setting the set_transition_state for repair. Can't we do the same for tablet_move?

```
--- a/service/storage_service.cc
+++ b/service/storage_service.cc
@@ -6198,9 +6198,7 @@ future<service::group0_guard> storage_service::wait_for_tablet_migration() {
 future<bool> storage_service::exec_tablet_migration(service::group0_guard guard, std::vector<canonical_mutation> updates, sstring reason) {
     rtlogger.info("{}", reason);
     rtlogger.trace("do update {} reason {}", updates, reason);
-    updates.emplace_back(topology_mutation_builder(guard.write_timestamp())
-        .set_transition_state(topology::transition_state::tablet_migration)
-        .build());
+    updates.emplace_back(topology_mutation_builder(guard.write_timestamp()).build());
     topology_change change{std::move(updates)};
     group0_command g0_cmd = _group0->client().prepare_command(std::move(change), guard, reason);
     try {
```

Reply:
> Can't we do the same for tablet_move?

No, because tablet move puts the tablet into transition, which is a state which should be bounded within tablet_migration/tablet_drain. There are topology ops which assume that tablets are not in transition when they run, and those ops may start running unless we set the topology transition state to "tablet_migration".